### PR TITLE
Habitat 2.0 reach_state task

### PR DESCRIPTION
**Pull request by vincentpierre**
## Motivation and Context

Before merging hab_suite_dev to main, this PR merges the minimal changes needed to get the reach_state task into main.

## How Has This Been Tested

Unit testing. Added a unit test testing the reach task for 20 updates and checking the reward is above 5.

### CODE REVIEW

File: habitat/tasks/rearrange/rearrange_grasp_manager.py [hunk 1]
```diff
@@ -228,6 +227,9 @@ def snap_to_obj(self, snap_obj_id: int, force: bool = True) -> None:
             )
 
         self._snapped_obj_id = snap_obj_id
+        if force:
+            # Set the transformation to be in the robot's hand already.
+            self.snap_rigid_obj.transformation = self._sim.robot.ee_transform
```
Comment (aclegg3) [inline habitat/tasks/rearrange/rearrange_grasp_manager.py:232]: Seems redundant

File: habitat/tasks/rearrange/rearrange_sim.py [hunk 2]
```diff
@@ -181,7 +195,7 @@ def reconfigure(self, config: Config):
         self._clear_objects(should_add_objects)
 
         self.prev_scene_id = ep_info["scene_id"]
-        self._viz_templates: Dict[float, int] = {}
+        self._viz_templates = {}
```
Comment (aclegg3) [inline habitat/tasks/rearrange/rearrange_sim.py:198]: Removing typing on purpose?

File: habitat/tasks/rearrange/rearrange_sim.py [hunk 3]
```diff
@@ -517,32 +565,20 @@ def step(self, action: Union[str, int]) -> Observations:
...
-        position: np.ndarray,
-        viz_id: Optional[int] = None,
-        r: float = 0.05,
-    ) -> int:
-        """Adds the sphere object to the specified position for visualization purpose.
-
-        :param position: global position of the visual sphere
-        :param viz_id: provided if moving an existing visual sphere instead of creating a new one
-        :param r: radius of the visual sphere
+    def visualize_position(self, position, viz_id=None, r=0.05) -> int:
```
Comment (aclegg3) [inline habitat/tasks/rearrange/rearrange_sim.py:568]: Removed the typing here on purpose?

File: habitat/tasks/rearrange/sub_tasks/reach_sensors.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from habitat.core.embodied_task import Measure
+from habitat.core.registry import registry
+from habitat.tasks.rearrange.rearrange_sensors import EndEffectorToRestDistance
+
+
+@registry.register_measure
+class RearrangeReachReward(Measure):
```
Comment (aclegg3) [inline habitat/tasks/rearrange/sub_tasks/reach_sensors.py:13]: Comments/typing may be helpful throughout

**Review by aclegg3 — COMMENTED** [review]
Quick review over the sim/task python. Didn't look at yamls or training code. Seeing some merge collisions may have removed typing or created some redundant code.

File: habitat/tasks/rearrange/rearrange_grasp_manager.py (hunk 1 above)
Comment (ASzot) [reply habitat/tasks/rearrange/rearrange_grasp_manager.py:232]: No, this is important to keep since some functionality is needed to automatically snap the object into the robot's hand, even if the object is far away.

File: habitat/tasks/rearrange/rearrange_sim.py [hunk 5]
```diff
@@ -261,6 +277,23 @@ def reconfigure(self, config: Config):
                 ao: ao.joint_positions for ao in self.art_objs
             }
 
+    def get_nav_pos(self, pos):
```
Comment (ASzot) [inline habitat/tasks/rearrange/rearrange_sim.py:280]: All of these constants were tuned for the ReplicaCAD scenes to deal with isolated navmesh islands. Is this code still necessary even with the updated ReplicaCAD scenes? This code will fail for other assets.

File: habitat/tasks/rearrange/rearrange_sim.py (hunk 2 above)
Comment (vincentpierre) [reply habitat/tasks/rearrange/rearrange_sim.py:198]: The type is given line 72. Adding type again is redundant no?

File: habitat/tasks/rearrange/rearrange_sim.py [hunk 6]
```diff
@@ -456,23 +496,32 @@ def step(self, action: Union[str, int]) -> Observations:
 
         self._update_markers()
 
-        if self._should_render_debug:
+        if self._debug_render:
+            rom = self.get_rigid_object_manager()
             self._try_acquire_context()
-            for obj_handle, _ in self.ep_info["targets"].items():
+            # Don't draw bounding boxes over target objects.
+            for obj_handle, _ in self._targets.items():
```
Comment (ASzot) [inline habitat/tasks/rearrange/rearrange_sim.py:503]: Why not draw bounding boxes over target objects? Was this comment intentional? It is important to draw the bounding boxes to distuingish from target positions and real objects. This is not important for Pick but it is for other tasks.

File: habitat_baselines/common/tensorboard_utils.py [hunk 7]
```diff
@@ -10,6 +10,20 @@
...
+    import wandb
+except ImportError:
+    pass
+
+
+def get_writer(config, **kwargs):
+    if config.WRITER_TYPE == "tb":
+        return TensorboardWriter(config.TENSORBOARD_DIR, **kwargs)
+    elif config.WRITER_TYPE == "wb":
+        return WeightsAndBiasesWriter(config)
```
Comment (ASzot) [inline habitat_baselines/common/tensorboard_utils.py:23]: Watch out, I added this code on a previous commit and it is not very well tested.

File: habitat/tasks/rearrange/rearrange_sim.py (hunk 6 above)
Comment (vincentpierre) [reply habitat/tasks/rearrange/rearrange_sim.py:503]: I think you are the one who added this comment : https://github.com/facebookresearch/habitat-lab/blame/hab_suite_dev/habitat/tasks/rearrange/rearrange_sim.py#L502 I am just porting it to main. If this code needs to change, can you advise?

File: habitat/tasks/rearrange/rearrange_grasp_manager.py (hunk 1 above)
Comment (vincentpierre) [reply habitat/tasks/rearrange/rearrange_grasp_manager.py]: It is my fault. I went too ast and did not realize the block 
```
        if force:
            # Set the transformation to be in the robot's hand already.
            self.snap_rigid_obj.transformation = self._sim.robot.ee_transform
```
was added twice (once on main and once on hab_suite_dev) resulting in literal redundant/duplicated code.

File: habitat_baselines/common/tensorboard_utils.py (hunk 7 above)
Comment (vincentpierre) [reply habitat_baselines/common/tensorboard_utils.py]: Ok, removed it since it is not necessary for this PR

File: habitat_baselines/rl/ddppo/policy/resnet_policy.py [hunk 8]
```diff
@@ -216,18 +224,26 @@ def __init__(
...
+            num_actions = get_num_actions(action_space)
+            self.prev_action_embedding = nn.Linear(num_actions, 32)
 
         self._n_prev_action = 32
         rnn_input_size = self._n_prev_action
 
+        self._fuse_keys = fuse_keys
+        if self._fuse_keys is not None:
+            rnn_input_size += sum(
+                [observation_space.spaces[k].shape[0] for k in self._fuse_keys]
```
Comment (Skylion007) [inline habitat_baselines/rl/ddppo/policy/resnet_policy.py:244]: Could be a generator expression
```suggestion
                (observation_space.spaces[k].shape[0] for k in self._fuse_keys)
```

File: habitat_baselines/utils/env_utils.py [hunk 9]
```diff
@@ -58,28 +58,29 @@ def construct_envs(
     if "*" in config.TASK_CONFIG.DATASET.CONTENT_SCENES:
         scenes = dataset.get_scenes_to_load(config.TASK_CONFIG.DATASET)
 
-    if num_environments > 1:
-        if len(scenes) == 0:
-            raise RuntimeError(
-                "No scenes to load, multiple process logic relies on being able to split scenes uniquely between processes"
-            )
+    if num_environments < 1:
+        raise RuntimeError("NUM_ENVIRONMENTS must be strictly positive")
```
Comment (Skylion007) [inline habitat_baselines/utils/env_utils.py:62]:
```suggestion
        raise RuntimeError(f"NUM_ENVIRONMENTS must be strictly positive: {num_environments}")
```

File: habitat_baselines/rl/ddppo/policy/resnet_policy.py [hunk 10]
```diff
@@ -107,17 +115,16 @@ def __init__(
...
 
-        if "depth" in observation_space.spaces:
-            self._n_input_depth = observation_space.spaces["depth"].shape[2]
-            spatial_size = observation_space.spaces["depth"].shape[0] // 2
-        else:
-            self._n_input_depth = 0
+        # Count total # of channels for rgb and for depth
+        self._n_input_rgb, self._n_input_depth = [
+            # sum() returns 0 for an empty list
+            sum([observation_space.spaces[k].shape[2] for k in keys])
```
Comment (Skylion007) [inline habitat_baselines/rl/ddppo/policy/resnet_policy.py:125]:
```suggestion
            sum((observation_space.spaces[k].shape[2] for k in keys))
```

File: habitat/tasks/rearrange/rearrange_sim.py [hunk 11]
```diff
@@ -360,22 +375,30 @@ def _add_objs(self, ep_info: Config, should_add_objects: bool) -> None:
...
+            new_target_handle = (
+                target_handle.split("_:")[0] + ".object_config.json"
+            )
+            matching_templates = (
+                obj_attr_mgr.get_templates_by_handle_substring(
+                    new_target_handle
+                )
+            )
+            ro = rom.add_object_by_template_handle(
+                list(matching_templates.keys())[0]
```
Comment (Skylion007) [inline habitat/tasks/rearrange/rearrange_sim.py:397]:
```suggestion
                next(iter(matching_templates.keys()))
```
^ Has O(n) complexity

File: habitat/tasks/rearrange/rearrange_sim.py [hunk 12]
```diff
@@ -161,7 +163,15 @@ def reconfigure(self, config: Config):
         self.ep_info = ep_info
         self._try_acquire_context()
 
-        if self.prev_scene_id != ep_info["scene_id"]:
+        self._targets = {}
+        for target_handle, transform in self.ep_info["targets"].items():
+            self._targets[target_handle] = mn.Matrix4(
```
Comment (Skylion007) [inline habitat/tasks/rearrange/rearrange_sim.py:168]: Do we have any assert here that transform[i,j] is of len(4)? Maybe consider adding it.

**Review by Skylion007 — COMMENTED** [review]
Added some nits.

File: habitat/tasks/rearrange/rearrange_sim.py (hunk 12 above)
Comment (vincentpierre) [reply habitat/tasks/rearrange/rearrange_sim.py:175]: Good catch, the error would have been very obscure to a new user. I added some asserts with a message
